File: 2014 Southeast USA Regionals Division 1/wordladder2.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <limits.h>

#define MAX_N 1000
#define MAX_LEN 20
#define INF 1000000000

typedef struct {
    int first, second;
} ti2;

typedef struct {
    int first, second, third;
} ti3;

int N;
char words[MAX_N][MAX_LEN];
int D[MAX_N][MAX_N];

int word_dist(int a, int b) {
    int res = 0;
    int len = strlen(words[a]);
    for (int i = 0; i < len; ++i) {
        if (words[a][i] != words[b][i]) ++res;
    }
    return res;
}

void diffs(int a, int b, char res[2][MAX_LEN]) {
    int pos[2], idx = 0;
    int len = strlen(words[a]);
    for (int i = 0; i < len; ++i) {
        if (words[a][i] != words[b][i]) {
            pos[idx++] = i;
        }
    }
    
    strcpy(res[0], words[a]);
    res[0][pos[0]] = words[b][pos[0]];
    
    strcpy(res[1], words[a]);
    res[1][pos[1]] = words[b][pos[1]];
}

char* solve(ti2* pairs, int pairs_count) {
    static char res[MAX_LEN];
    res[0] = '\0';
    
    for (int i = 0; i < pairs_count; ++i) {
        int a = pairs[i].first;
        int b = pairs[i].second;
        char temp[2][MAX_LEN];
        diffs(a, b, temp);
        
        if (res[0] == '\0') {
            strcpy(res, temp[0]);
        }
        if (strcmp(temp[0], res) < 0) {
            strcpy(res, temp[0]);
        }
        if (strcmp(temp[1], res) < 0) {
            strcpy(res, temp[1]);
        }
    }
    return res;
}

int trace(int dist[][2], int cur, ti2* pairs, int* pairs_count) {
    *pairs_count = 0;
    int queue[MAX_N];
    int front = 0, rear = 0;
    queue[rear++] = cur;
    
    while (front < rear) {
        cur = queue[front++];
        for (int i = 0; i < N; ++i) {
            int d = D[i][cur];
            if (d > 2) continue;
            if (d == 2 && dist[i][0] + d == dist[cur][1]) {
                pairs[*pairs_count].first = i;
                pairs[*pairs_count].second = cur;
                (*pairs_count)++;
            }
            if (d == 1 && dist[i][1] + d == dist[cur][1]) {
                queue[rear++] = i;
            }
        }
    }
    return *pairs_count;
}

int dijkstra(int src, int dest, ti2* pairs, int* pairs_count) {
    int dist[MAX_N][2];
    for (int i = 0; i < N; ++i) {
        dist[i][0] = INF;
        dist[i][1] = INF;
    }
    
    // Simple queue simulation for priority queue
    ti3 queue[MAX_N * 2];
    int size = 0;
    
    dist[src][0] = 0;
    queue[size].first = 0;
    queue[size].second = 0;
    queue[size].third = src;
    size++;
    
    while (size > 0) {
        // Find min element (simple min search)
        int min_idx = 0;
        for (int i = 1; i < size; ++i) {
            if (queue[i].first < queue[min_idx].first) {
                min_idx = i;
            }
        }
        
        int s = queue[min_idx].first;
        int k = queue[min_idx].second;
        int cur = queue[min_idx].third;
        
        // Remove element
        for (int i = min_idx; i < size - 1; ++i) {
            queue[i] = queue[i + 1];
        }
        size--;
        
        for (int nxt = 0; nxt < N; ++nxt) {
            int d = D[cur][nxt];
            int nk = (d == 2);
            if (d > 2 || (nk && k)) continue;
            if (k) nk = 1;
            
            if (s + d < dist[nxt][nk]) {
                dist[nxt][nk] = s + d;
                queue[size].first = s + d;
                queue[size].second = (d == 2) ? 1 : k;
                queue[size].third = nxt;
                size++;
            }
        }
    }
    
    if (dist[dest][0] <= dist[dest][1]) {
        return dist[dest][0];
    }
    
    *pairs_count = trace(dist, dest, pairs, pairs_count);
    return dist[dest][1];
}

int main() {
    scanf("%d", &N);
    for (int i = 0; i < N; ++i) {
        scanf("%s", words[i]);
    }
    
    for (int i = 0; i < N; ++i) {
        for (int j = 0; j < N; ++j) {
            D[i][j] = INF;
        }
    }
    
    for (int i = 0; i < N; ++i) {
        for (int j = i + 1; j < N; ++j) {
            D[i][j] = D[j][i] = word_dist(i, j);
        }
    }
    
    ti2 pairs[MAX_N];
    int pairs_count;
    int steps = dijkstra(0, 1, pairs, &pairs_count);
    
    if (steps == INF || pairs_count == 0) {
        printf("0\n");
    } else {
        printf("%s\n", solve(pairs, pairs_count));
    }
    printf("%d\n", (steps == INF) ? -1 : steps);
    
    return 0;
}

File: 2014 Southeast USA Regionals Division 1/wordladder2.py
import heapq
import sys

INF = 10**9

def word_dist(a, b, words):
    return sum(1 for i in range(len(words[a])) if words[a][i] != words[b][i])

def diffs(a, b, words):
    pos = [i for i in range(len(words[a])) if words[a][i] != words[b][i]]
    res1 = list(words[a])
    res2 = list(words[a])
    res1[pos[0]] = words[b][pos[0]]
    res2[pos[1]] = words[b][pos[1]]
    return [''.join(res1), ''.join(res2)]

def solve(pairs, words):
    res = ""
    for a, b in pairs:
        R = diffs(a, b, words)
        if not res:
            res = R[0]
        res = min(res, R[0], R[1])
    return res

def trace(dist, cur, D, N):
    res = []
    Q = [cur]
    while Q:
        cur = Q.pop(0)
        for i in range(N):
            d = D[i][cur]
            if d > 2:
                continue
            if d == 2 and dist[i][0] + d == dist[cur][1]:
                res.append((i, cur))
            if d == 1 and dist[i][1] + d == dist[cur][1]:
                Q.append(i)
    return res

def dijkstra(src, dest, words, D, N):
    dist = [[INF, INF] for _ in range(N)]
    heap = []
    
    dist[src][0] = 0
    heapq.heappush(heap, (0, 0, src))
    
    while heap:
        s, k, cur = heapq.heappop(heap)
        if s > dist[cur][k]:
            continue
            
        for nxt in range(N):
            d = D[cur][nxt]
            nk = 1 if d == 2 else 0
            if d > 2 or (nk == 1 and k == 1):
                continue
            if k == 1:
                nk = 1
                
            if s + d < dist[nxt][nk]:
                dist[nxt][nk] = s + d
                heapq.heappush(heap, (s + d, nk, nxt))
    
    if dist[dest][0] <= dist[dest][1]:
        return dist[dest][0], []
    
    pairs = trace(dist, dest, D, N)
    return dist[dest][1], pairs

def main():
    data = sys.stdin.read().splitlines()
    if not data:
        return
    
    N = int(data[0])
    words = [line.strip() for line in data[1:1+N]]
    
    D = [[INF] * N for _ in range(N)]
    for i in range(N):
        for j in range(i+1, N):
            dist = word_dist(i, j, words)
            D[i][j] = D[j][i] = dist
    
    steps, pairs = dijkstra(0, 1, words, D, N)
    
    if steps == INF or not pairs:
        print(0)
    else:
        print(solve(pairs, words))
    print(-1 if steps == INF else steps)

if __name__ == "__main__":
    main()